### DISCUSSION TITLE
Bump Cask/Formula PR for Intel and M1

### DISCUSSION THREAD

**Discussion by cesarcoatl** [Homebrew/discussions#561] [Everyday usage]
I assume that a simple `brew bump-[cask|formula]-pr` won't work for casks or formulae that have different download links; one for Intel Macs, and one for M1.

For cask-versions that may release on a daily basis a manual PR will have to wait for approval from a maintainer instead of being automatically approved by BrewTestBot. Or at least that's my idea.

Is there something in the works or maybe something already in place for "bumping" versions for casks or formulae that fit this scenario?

Thanks

**Comment by SeekingMeaning** [comment]
Edit: See https://github.com/Homebrew/brew/pull/10345

This may require new syntax similar to the `language` stanza in casks (e.g. [`firefox`](https://github.com/Homebrew/homebrew-cask/blob/master/Casks/firefox.rb))

**Reply by SeekingMeaning** [reply]
Maybe something like this:

```rb
cask "macintoshjs" do
  version "1.1.0"

  cpu :intel do
    sha256 "4ca41517f15c594e718da622074a2160754cb8f1293cb1fad908f6d0ae585384"
    "x64"
  end

  cpu :arm do
    sha256 "0af95bbcc075f939d6c3e1fdcdab0d8da2760e5546aecfbc82767e326640740c"
    "arm64"
  end

  url "https://github.com/felixrieseberg/macintosh.js/releases/download/v#{version}/macintosh.js-darwin-#{cpu}-#{version}.zip"
  # ...
end
```

**Reply by vitorgalvao** [reply]
The issue is that `version` is sometimes different for Intel and ARM. See https://github.com/Homebrew/homebrew-cask-versions/pull/10205/files for an example.

**Reply by cesarcoatl** [reply]
My question is not so much about cask syntax but about how one would use `brew bump-cask-pr`.

Currently we can pass the version separated by a comma, but I don't know if that would be the same for the SHA-256 hash.
 
```bash
$ brew bump-cask-pr --version #{version.before_comma},#{version.after_comma} --sha256 #{hash.before_comma},#{hash.after_comma} foo
```

Otherwise all version bumps would be submitted and approved manually rather than automatically.

**Reply by vitorgalvao** [reply]
> Currently we can pass the version separated by a comma

That’s for a single version. It’s not changing both.

> but I don't know if that would be the same for the SHA-256 hash.

There’s no need to manually pass it, it’s determined automatically.

**Reply by SeekingMeaning** [reply]
> The issue is that `version` is sometimes different for Intel and ARM. See https://github.com/Homebrew/homebrew-cask-versions/pull/10205/files for an example.

This works, albeit somewhat hacky: (Edit: we could fix the `cpu`-call hack by immediately executing the block if the CPU matches)

```rb
cask "visual-studio-code-insiders" do
  cpu :intel do
    version "1.53.0,6dc779565e1e4a26104002fde340b68c157c3bc5"
    sha256 "ea08cd72f7501223d1c888b300bd961df27e80558e4e51c868740f651f3912ce"
    ""
  end

  cpu :arm do
    version "1.53.0,e17aea136d32fdb06d9b32184b04ccd36bb51e9c"
    sha256 "3f9bdb6867636cc9116d96a09aef53a927e1636609df33bfe23a1275107a04b6"
    "-arm64"
  end

  cpu # Need to call `cpu` method to initialize `version`

  url "https://az764295.vo.msecnd.net/insider/#{version.after_comma}/VSCode-darwin#{cpu}.zip",
      verified: "az764295.vo.msecnd.net/insider/"
  # ...
end
```

**Reply by SeekingMeaning** [reply]
> My question is not so much about cask syntax but about how one would use `brew bump-cask-pr`.

Yes, this new syntax allows us to determine what CPUs are supported and then use that to determine which version numbers and sha256 hashes need to be updated

**Reply by cesarcoatl** [reply]
> There’s no need to manually pass it, it’s determined automatically.

I had forgotten about that, since I got used to passing it myself.

> That’s for a single version. It’s not changing both.

Also true. Although in the case for the latest releases of [visual-studio-code-insiders](https://github.com/Homebrew/homebrew-cask-versions/blob/bd046e656966f6552e6939dba8eb950f79bd45c2/Casks/visual-studio-code-insiders.rb) the versions were the same for both architectures, but the Cask had to be slightly altered.

I'll look into the new syntax, as it seems that this "releases for both architectures" can be solved in the Cask definition.

Thanks.

**Comment by cesarcoatl** [comment]
Additional findings.

There's a new release for `visual-studio-code-insiders`, and in my case only the SHA-256 hash for Intel is being updated with a simple `brew bump-cask-pr`, because my machine is Intel. So I assume that the cask audit would fail because the SHA-256 hash for M1 would not be updated.

And I would expect if I were to run `brew bum-cask-pr` from a Mac with M1, only the M1 info would be updated.

So, back to the original question.

How would one got about creating bump PRs for both architectures via `brew bump-cask-pr`?

See console output:

```bash
$ brew bump-cask-pr --dry-run --debug --verbose --version 1.53.0,eaa959d34b863dec932077d94daece22504a396b visual-studio-code-insiders
git config --replace-all homebrew.devcmdrun true
/usr/bin/curl --disable --globoff --show-error --user-agent Homebrew/2.7.5-39-ge8667b6\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 10.15.7\)\ curl/7.64.1 --header Accept-Language:\ en --retry 3 --location https://api.github.com/search/issues\?q=visual-studio-code-insiders\+repo\%3AHomebrew\%2Fhomebrew-cask-versions\+state\%3Aopen\+in\%3Atitle\&per_page=100 --header Accept:\ application/vnd.github.v3\+json --write-out '
'\%\{http_code\} --header Accept:\ application/vnd.github.antiope-preview\+json --header Authorization:\ token\ ****** --dump-header /private/tmp/github_api_headers20210118-5063-1p29flh
/usr/bin/curl --disable --globoff --show-error --user-agent Homebrew/2.7.5-39-ge8667b6\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 10.15.7\)\ curl/7.64.1 --header Accept-Language:\ en --retry 3 --location --silent --head --request GET https://az764295.vo.msecnd.net/insider/eaa959d34b863dec932077d94daece22504a396b/VSCode-darwin.zip
==> Downloading https://az764295.vo.msecnd.net/insider/eaa959d34b863dec932077d94daece22504a396b/VSCode-darwin.zip
/usr/bin/curl --disable --globoff --show-error --user-agent Homebrew/2.7.5-39-ge8667b6\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 10.15.7\)\ curl/7.64.1 --header Accept-Language:\ en --retry 3 --location --range 0-1 --dump-header - --write-out \%\{http_code\} --output /dev/null https://az764295.vo.msecnd.net/insider/eaa959d34b863dec932077d94daece22504a396b/VSCode-darwin.zip
/usr/bin/curl --disable --globoff --show-error --user-agent Homebrew/2.7.5-39-ge8667b6\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 10.15.7\)\ curl/7.64.1 --header Accept-Language:\ en --fail --retry 3 --location --remote-time --continue-at 0 --output /Users/thecesrom/Library/Caches/Homebrew/downloads/e624e9214433234cd554f44a065bbd91ee29fd59d9e78d7dd4eb0caa0948a47f--VSCode-darwin.zip.incomplete https://az764295.vo.msecnd.net/insider/eaa959d34b863dec932077d94daece22504a396b/VSCode-darwin.zip
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 97.1M  100 97.1M    0     0  19.3M      0  0:00:05  0:00:05 --:--:-- 20.1M
==> Verifying checksum for 'e624e9214433234cd554f44a065bbd91ee29fd59d9e78d7dd4eb0caa0948a47f--VSCode-darwin.zip'.
Warning: Cannot verify integrity of 'e624e9214433234cd554f44a065bbd91ee29fd59d9e78d7dd4eb0caa0948a47f--VSCode-darwin.zip'.
No checksum was provided for this resource.
For your reference, the checksum is:
  sha256 "8a0090341f35b61bb879242b3a7517467a46ed46943b8811c7737c349c135bcf"
==> replace /version\s+["']1\.53\.0,1a6eef3170be2a11751ee50d450328d458f7bfac["']/m with "version \"1.53.0,eaa959d34b863dec932077d94daece22504a396b\""
==> replace /sha256\s+["']a1b2ddc719b1201e63c63268e13a9528e56779f63670120b393e73710022f56a["']/m with "sha256 \"8a0090341f35b61bb879242b3a7517467a46ed46943b8811c7737c349c135bcf\""
==> brew audit --cask visual-studio-code-insiders.rb
==> brew style --fix visual-studio-code-insiders.rb
/usr/local/Homebrew/Library/Homebrew/shims/scm/git --version
==> try to fork repository with GitHub API
==> git add /usr/local/Homebrew/Library/Taps/homebrew/homebrew-cask-versions/Casks/visual-studio-code-insiders.rb
==> git checkout --no-track -b bump-visual-studio-code-insiders-1.53.0-eaa959d34b863dec932077d94daece22504a396b origin/master
==> git commit --no-edit --verbose --message='Update visual-studio-code-insiders from 1.53.0,1a6eef3170be2a11751ee50d450328d458f7bfac to 1.53.0,eaa959d34b863dec932077d94daece22504a396b' -- /usr/local/Homebrew/Library/Taps/homebrew/homebrew-cask-versions/Casks/visual-studio-code-insiders.rb
==> git push --set-upstream $HUB_REMOTE bump-visual-studio-code-insiders-1.53.0-eaa959d34b863dec932077d94daece22504a396b:bump-visual-studio-code-insiders-1.53.0-eaa959d34b863dec932077d94daece22504a396b
==> git checkout --quiet -
==> create pull request with GitHub API (base branch: master)
```

**Reply by SeekingMeaning** [reply]
A workaround that should work is to manually update the SHA-256 for ARM only and then run `brew bump-cask-pr`, which will update the SHA-256 for Intel and then automatically submit the PR with both SHA-256 checksums updated

**Reply by cesarcoatl** [reply]
I resorted into doing it manually within GitHub, instead of doing it from the Terminal. [Homebrew/homebrew-cask-versions#10230](https://github.com/Homebrew/homebrew-cask-versions/pull/10230).

Since this PR does not alter the Cask, only version and hash, apparently it is a good candidate for automerge.

But I do miss how easy was to accomplish this with just one command.

**Comment by tabris233** [comment]
Excuse me, what is the progress of this discussion?

**Reply by cesarcoatl** [reply]
`brew bump-cask-pr` updates the `sha256` stanza accordingly.

My latest PR https://github.com/Homebrew/homebrew-cask/pull/120322 was created on an Intel Mac just by running the following command:

```bash
$ brew bump-cask-pr --version 1.65.2 visual-studio-code
==> Downloading https://update.code.visualstudio.com/1.65.2/darwin/stable
==> Downloading from https://az764295.vo.msecnd.net/stable/c722ca6c7eed3d7987c0d5c3df5c45f6b15e77d1/VSCode-darwin.zip
######################################################################## 100.0%
==> Downloading https://update.code.visualstudio.com/1.65.2/darwin-arm64/stable
==> Downloading from https://az764295.vo.msecnd.net/stable/c722ca6c7eed3d7987c0d5c3df5c45f6b15e77d1/VSCode-darwin-arm64.
######################################################################## 100.0%
==> replace /version\s+["']1\.65\.1["']/m with "version \"1.65.2\""
==> replace "4019d040ae7fdf62908e78570efa06eb6f76237f8b5f0959295d6ee0b1a5efcc" with "a351e27b073eb153a5fb51a3dab7a2e3188
==> replace /sha256\s+["']e56dfdb3b66a5640050df9aa2eec1a90649e028c8e25eb72be19f3409f4032f2["']/m with "sha256 \"6202e1ae
audit for visual-studio-code: passed

1 file inspected, no offenses detected
```

And the PR was created.
